File: 383.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        // time: O(m + n), space: O(1)
        Map<Character, Integer> m = new HashMap<>();
        for (char c : magazine.toCharArray()) {
            m.put(c, m.getOrDefault(c, 0) + 1);
        }
        for (char c : ransomNote.toCharArray()) {
            if (!m.containsKey(c) || m.get(c) <= 0) {
                return false;
            }
            m.put(c, m.get(c) - 1);
        }
        return true;
    }
}

File: 383.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # time: O(m + n), space: O(1)
        count = {}
        for c in magazine:
            count[c] = count.get(c, 0) + 1
        for c in ransomNote:
            if c not in count:
                return False
            count[c] -= 1
            if count[c] < 0:
                return False
        return True